feat(Sales Invoice): Convert the PDF data to PDF/A-3 using ghostscript.

timschneider opened the pull request:
Convert PDF to PDF/A-3 if ghostscript is installed.
Fail on error and do not output pdf/a-3 data with xml data, erpnext will output the default pdf.

The switch `"-dPDFACompatibilityPolicy=1",` will behave like
```

PDFACompatibilityPolicy integer

    When an operation (e.g. pdfmark) is encountered which cannot be emitted in a PDF/A compliant file, this policy is consulted, there are currently three possible values:

    0 - (default) Include the feature or operation in the output file, the file will not be PDF/A compliant. Because the document Catalog is emitted before this is encountered, the file will still contain PDF/A metadata but will not be compliant. A warning will be emitted in this case.

    1 - The feature or operation is ignored, the resulting PDF file will be PDF/A compliant. A warning will be emitted for every elided feature.

    2 - Processing of the file is aborted with an error, the exact error may vary depending on the nature of the PDF/A incompatibility.

```

Maybe change it later to 2 or make it configurable.

timschneider committed:
Convert the PDF data to PDF/A-3 using ghostscript.
Fail on error and do not output pdf/a-3 data with
xml data.

Convert PDF to PDF/A-3 if ghostscript is installed

timschneider committed:
refactor annotation

0xD0M1M0 commented:
Does Ghostscript needs to be added to pyprojects.toml so it can be installed as dependency?

barredterra commented on eu_einvoice/european_e_invoice/custom/sales_invoice.py:
```diff
@@ -781,6 +781,73 @@ def download_pdf(
+	if lib_path:
+		icc_path = os.path.join(lib_path, "iccprofiles/srgb.icc")
+		if os.path.isfile(icc_path):
+			os.symlink(icc_path, "srgb.icc")
```
I'm not sure if we are (or should be) allowed to do this. An ERPNext extension writing around in the host system (apart from designated folders) seems bad.

I don't know much about containers, but I have a hunch that this location could be read-only in such environments.

timschneider replied:
I can understand your point and I've the same feeling about it! The symbolic link will end in a docker environment in `/home/frappe/frappe-bench/sites/srgb.icc` 

The problem is, that the srgb.icc is linked in the PDFA_def.ps relative to $PWD e.g. in the current folder.

I can think of the following different solutions

1. include the PDFA_def.ps and the srgb.icc in the package, but the license of ghostscript and the icc profile and so one is cumbersome. So I think it is better not to include any of the ghostscript stuff
2. copy the PDFA_def.ps file to a temporary folder and modify it to include the absolute path of the icc profile
3. change the current working directory of ghostscript to the icc profiles folder

I think the best option is to change the cwd of the subprocess call to gs to make it the icc profiles path.

timschneider committed:
reworked: change cwd to icc profiles folder instead of creating a symbolic link inside the current folder.

timschneider commented:
@0xD0M1M0 
> Does Ghostscript needs to be added to pyprojects.toml so it can be installed as dependency?

I don't know if it is possible to add an external dependency to the pyproject.toml that will be installed with the plugin.

timschneider committed:
refactor: arrow down the regex

timschneider committed:
refactor

timschneider committed:
- Redirect Ghostscript error output to stderr.
- renamed path to library_path
- specified comment

barredterra commented on eu_einvoice/european_e_invoice/custom/sales_invoice.py:
```diff
@@ -794,6 +874,9 @@ def attach_xml_to_pdf(invoice_id: str, pdf_data: bytes) -> bytes:
 	from drafthorse.pdf import attach_xml
 
+	if _is_ghostscript_installed():
+		pdf_data = _convert_pdf_to_pdfa(pdf_data)
```
```suggestion
		try:
			pdf_data = _convert_pdf_to_pdfa(pdf_data)
		except RuntimeError:
			frappe.log_error()
```

This way we can continue to write invoices when Ghostscript is installed but fails, while at the same time alerting the System Manager that something is off.

timschneider committed:
Refactor Ghostscript error handling in attach_xml_to_pdf function as suggested by @barredterra

timschneider committed:
Refactor Ghostscript error handling in attach_xml_to_pdf function

2 - Processing of the file is aborted with an error, the exact error may vary depending on the nature of the PDF/A incompatibility.

It will fallback to a normal pdf with embedded xml data and log the error in the frappe log.

timschneider committed:
Removed unused import re from _convert_pdf_to_pdfa

barredterra commented:
@Mergifyio backport version-14 develop
